src/Pos/Client/Txp/Util.hs

```diff
@@ -206,60 +260,184 @@ prepareInpsOuts utxo addrs outputs = do
+    => Utxo
+    -> RedeemSecretKey
+    -> TxOutputs
+    -> m (Either TxError TxAux)
+createRedemptionTx utxo rsk outputs = runExceptT $ do
+    TxRaw {..} <- prepareTxRaw utxo outputs (TxFee $ mkCoin 0)
+    let bareInputs = snd <$> trInputs
+    properOutputs <- overrideTxDistrBoot trOutputs
+    pure $ makeRedemptionTx rsk bareInputs properOutputs
+
+-----------------------------------------------------------------------------
+-- Fees logic
+-----------------------------------------------------------------------------
+
+prepareTxWithFee
+    :: TxDistrMode ctx m
+    => Utxo
+    -> TxOutputs
+    -> ExceptT TxError m TxRaw
+prepareTxWithFee utxo outputs = do
+    feePolicy <- bvdTxFeePolicy <$> lift gsAdoptedBVData
+    case feePolicy of
+        TxFeePolicyUnknown w _ -> throwTxError $
+            sformat ("Unknown fee policy, tag: "%build) w
+        TxFeePolicyTxSizeLinear linearPolicy ->
+            stabilizeTxFee linearPolicy utxo outputs
+
+-- Read-only function.
+computeTxFee
+    :: TxDistrMode ctx m
+    => Utxo
+    -> TxOutputs
+    -> ExceptT TxError m TxFee
+computeTxFee utxo outputs = do
+    feePolicy <- bvdTxFeePolicy <$> gsAdoptedBVData
+    case feePolicy of
+        TxFeePolicyUnknown w _ -> throwTxError $
+            sformat ("Unknown fee policy, tag: "%build) w
+        TxFeePolicyTxSizeLinear linearPolicy -> do
+            txAux <- createFakeTxFromRawTx <$>
+                     stabilizeTxFee linearPolicy utxo outputs
+            txToLinearFee linearPolicy txAux
+
+-- | Search such spendings that transaction's fee would be stable.
+stabilizeTxFee
+    :: forall m. MonadError TxError m
+    => TxSizeLinear
+    -> Utxo
+    -> TxOutputs
+    -> m TxRaw
+stabilizeTxFee linearPolicy utxo outputs =
+    stabilizeTxFeeDo 5 (TxFee $ mkCoin 0)
+  where
+    stabilizeTxFeeDo :: Int -> TxFee -> m TxRaw
+    stabilizeTxFeeDo 0 _ = throwTxError "Couldn't stabilize tx fee after 5 attempts"
+    stabilizeTxFeeDo attempt expectedFee = do
+        txRaw <- prepareTxRaw utxo outputs expectedFee
+        txFee <- txToLinearFee linearPolicy $
+                 createFakeTxFromRawTx txRaw
+        if expectedFee == txFee
+            then pure txRaw
+            else stabilizeTxFeeDo (attempt - 1) txFee
+
+txToLinearFee
+    :: MonadError TxError m
+    => TxSizeLinear -> TxAux -> m TxFee
+txToLinearFee linearPolicy =
+    either throwError pure .
+    bimap invalidFee TxFee .
+    integerToCoin .
+    ceiling .
+    calculateTxSizeLinear linearPolicy .
+    biSize @TxAux
+  where
+    invalidFee reason = TxError ("Invalid fee: " <> reason)
+
+createFakeTxFromRawTx :: TxRaw -> TxAux
+createFakeTxFromRawTx TxRaw{..} =
+    let fakeAddr = txOutAddress . toaOut . NE.head $ trOutputs
+        fakeOutMB
+            | trRemaining == mkCoin 0 = Nothing
+            | otherwise = Just $ TxOutAux (TxOut fakeAddr trRemaining) []
+        txOutsWithRem = maybe trOutputs (\remTx -> remTx :| toList trOutputs) fakeOutMB
```

Comment:
which function? you mean `mkOutputsWithRem`?
but it's monadic, and here we have pure context. Though it may be a good idea to abstract a pure part away.